ktor-network/common/src/io/ktor/network/selector/SocketBase.kt

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2014-2025 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.network.selector
+
+import io.ktor.network.sockets.*
+import io.ktor.utils.io.*
+import kotlinx.atomicfu.*
+import kotlinx.coroutines.*
+import kotlinx.io.*
+import kotlin.coroutines.*
+
+internal abstract class SocketBase(
+    parent: CoroutineContext
+) : ReadWriteSocket, SelectableBase(), CoroutineScope {
+
+    private val closeFlag = atomic(false)
+
+    internal val readerJob = atomic<ReaderJob?>(null)
+
+    internal val writerJob = atomic<WriterJob?>(null)
+
+    override val socketContext: CompletableJob = Job(parent[Job])
+
+    override val coroutineContext: CoroutineContext
+        get() = socketContext
+
+    override fun dispose() {
+        close()
+    }
+
+    override fun close() {
+        if (!closeFlag.compareAndSet(false, true)) return
+
+        readerJob.value?.channel?.close()
```

Comment:
This line was removed by https://github.com/ktorio/ktor/pull/4632/files on main branch, but this causes socket close to get stuck on JVM as well. For now I reverted it to get the test working, but please let me know what to do with this. Any ideas?